### ISSUE TITLE
Address rewrite / masquerade

### ISSUE THREAD

**Issue by kusuriya** [OpenSMTPD/OpenSMTPD#172]
Address rewriting with rules would be very helpful to have and useful for being feature competitive.

**Comment by poolpOrg** [comment]
We recently ran into a problem that forced us to implement very basic masquerading of the local domain to fix an issue when exchanging mails with other MTA.

The details of why and how are described on this blog post, I won't explain them again:
https://poolp.org/0xa86e/Some-OpenSMTPD-overview,-part-3

Long story short, to work around the issue the daemon **HAS** to perform some basic masquerading for the local domain, it has left the "feature" set to become a mandatory requirement and we don't want that in filters as we never want it disabled.

I implemented a mechanism to allow us to alter headers and a function to perform the basic masquerading we were required to do.

Now, with this code in, it no longer makes sense to move the "general" masquerading code to a filter since what we wanted to avoid having in the daemon has already been committed ...

I have started working on it, however I want #227 to be in first as it provides a bit of code I want to rely upon.

**Comment by jirutka** [comment]
This is great! I’m currently using a very simple (quite limited) filter to rewrite `From:` header. The problem is that filter is defined on `listen`, not `accept` (rule) directive, so I can’t simply use it only for specific rules.

It’d be great if your solution for header rewrite didn’t have this limitation.

```python
# coding=utf-8
import filter
from email.utils import formataddr
from email.utils import parseaddr

REWRITE_FROM = 'no-reply@example.org'

def on_dataline(id, line):
    if line.startswith('From:'):
        email = parseaddr(line)[1]

        if email != REWRITE_FROM:
            print("[rewrite filter] Rewriting %s to %s" % (email, REWRITE_FROM))
            line = line.replace(email, REWRITE_FROM)

    return filter.writeln(id, line)
```

**Comment by poolpOrg** [comment]
I can tell you straight ahead that it will have this limitation, here's why:

    $ nc localhost 25
    220 debug.poolp.org ESMTP OpenSMTPD
    HELO me
    250 debug.poolp.org Hello me [127.0.0.1], pleased to meet you
    MAIL FROM:<gilles@debug.poolp.org>
    250 2.0.0: Ok
    RCPT TO:<gilles@openbsd.org>
    250 2.1.5 Destination address valid: Recipient ok
    RCPT TO:<gilles@poolp.org> 
    250 2.1.5 Destination address valid: Recipient ok
    DATA
    354 Enter mail, end with "." on a line by itself
    From: gilles

    test
    .
    250 2.0.0: 538ecf02 Message accepted for delivery

Now, let's imagine that masquerading is a per-rule decision:

If my rule to relay to openbsd.org wants masquerading and my rule to relay to poolp.org does not want masquerading, my transaction suddenly faces a problem as it cannot create a unique copy of the message with different headers for the different recipients.

This is actually the exact same reason why the filters are per-listener and not per-rule, it would lead to problems if two rules are matched in the same transaction that do not want the same filter behaviour.

**Comment by poolpOrg** [comment]
this is finished in my sandbox, awaiting for ok's to commit

**Comment by poolpOrg** [comment]
Initial support for masquerading was committed yesterday to rewrite the content of the From header with the sender provided at the SMTP level.

I'm closing this ticket and I'll reopen one to discuss extending this support to rewriting From to a different sender than the SMTP level one.

**Comment by jirutka** [comment]
The last commit in this repo is 3 days old (d95780f8d287ef270f6174be3c7e3460fe9723c1) and no rewrite support in it…

**Comment by poolpOrg** [comment]
On Sat, Feb 14, 2015 at 07:40:18AM -0800, Jakub Jirutka wrote:
> The last commit in this repo is 3 days old (d95780f8d287ef270f6174be3c7e3460fe9723c1) and no rewrite support in it???
> 

Ooops,

I have forgot to push my master branch to our repository.

I'm currently working outside without remote access to my desktop, but I
will push the commit as soon as I get back home in a few hours.

-- 
Gilles Chehade

https://www.poolp.org                                          @poolpOrg

**Comment by athompso** [comment]
Gilles, you've indicated that #472, #474, and #476 are (or will be) solved by this issue (#172) somehow, but I am utterly at a loss to understand how.
I'm still accidentally bouncing legacy addresses that use "-" as a tag delimiter, and am still awaiting either a solution or someone to explain/document the solution for *inbound* mail.  The comments in this ticket imply that there exists some rewriting support, but I'm just not seeing how to apply it to the LHS on inbound email.
Am I missing something obvious?
(I'm running OpenSMTP as shipped in OpenBSD-5.9-RELEASE right now.)

**Comment by poolpOrg** [comment]
No, this is a misunderstanding.

What I said about #472 and #476 is that they are not going to be implemented in the daemon but that they can be solved using the same solution as what was planned for this ticket: a filter.

It turns out that this ticket required a daemon-side fix for a completely different reason that was preventing interop with other MTA and was causing undesirable side-effects. We added to the daemon the minimal changes required for interop, more fancy rewriting/masquerade will still be done outside the daemon in a filter.

This doesn't change the state of things regarding #472 and #476, we will provide this feature as part of a rewrite filter and not as part of the daemon.
